crates/bevy_animation/src/lib.rs

```diff
@@ -479,174 +485,217 @@ impl AnimationPlayer {
+
+        let paused = player.paused;
+        if paused {
+            continue;
+        }
+
+        // Update the main animation.
+        update_animation(&mut player.animation, paused, &time, &animation_clips);
+
+        // Update transition animations.
+        for transition in &mut player.transitions {
+            update_animation(&mut transition.animation, paused, &time, &animation_clips)
+        }
+    }
 }
 
-#[allow(clippy::too_many_arguments)]
-fn run_animation_player(
-    root: Entity,
-    mut player: Mut<AnimationPlayer>,
-    time: &Time,
-    animations: &Assets<AnimationClip>,
-    names: &Query<&Name>,
-    transforms: &Query<&mut Transform>,
-    morphs: &Query<&mut MorphWeights>,
-    maybe_parent: Option<&Parent>,
-    parents: &Query<(Has<AnimationPlayer>, Option<&Parent>)>,
-    children: &Query<&Children>,
+pub fn animate_targets(
+    world: &mut World,
+    mut system_state: Local<
+        Option<
+            SystemState<(
+                Query<&AnimationPlayer, Without<AnimationTarget>>,
+                Query<EntityMut, With<AnimationTarget>>,
+            )>,
+        >,
+    >,
 ) {
-    let paused = player.paused;
-    // Continue if paused unless the `AnimationPlayer` was changed
-    // This allow the animation to still be updated if the player.elapsed field was manually updated in pause
-    if paused && !player.is_changed() {
-        return;
+    if system_state.is_none() {
+        *system_state = Some(SystemState::new(world));
     }
 
-    // Apply the main animation
-    apply_animation(
-        1.0,
-        &mut player.animation,
-        paused,
-        root,
-        time,
-        animations,
-        names,
-        transforms,
-        morphs,
-        maybe_parent,
-        parents,
-        children,
-    );
-
-    // Apply any potential fade-out transitions from previous animations
-    for AnimationTransition {
-        current_weight,
-        animation,
-        ..
-    } in &mut player.transitions
-    {
-        apply_animation(
-            *current_weight,
-            animation,
-            paused,
-            root,
-            time,
-            animations,
-            names,
-            transforms,
-            morphs,
-            maybe_parent,
-            parents,
-            children,
-        );
-    }
+    // This elaborate setup using `resource_scope` is a workaround for the fact
+    // that `EntityMut` queries lock out resources. This is a known bug in Bevy.
+    world.resource_scope(|world, clips: Mut<Assets<AnimationClip>>| {
+        let (players, mut targets) = system_state.as_mut().unwrap().get_mut(world);
+        targets.par_iter_mut().for_each(|mut target_entity| {
+            let target = target_entity.get::<AnimationTarget>().unwrap();
+            let Ok(player) = players.get(target.root) else {
+                error!(
+                    "Couldn't find the animation player for {:?}",
+                    target_entity.id()
```

Comment:
nit: Since the animation system heavily relies on `Name`, it would be nice for debugging to print the `Name` of the entity.

Reply:
We can actually still do this here, if we add optional names to the targets query.